File: iosgl/iosgl/bt/event/EventDispatcher.m
//
//  EventDispatcher.m
//  iosgl
//
//  Created by zhao on 23/2/2020.
//  Copyright © 2020 zhao. All rights reserved.
//
#import "GL_Header.h"
#import "EventDispatcher.h"
#import "EventBindVo.h"
 //Object val, BaseEvent event
@implementation EventDispatcher
-(void)addEventListener:(NSString*)types  callback:(EventCallBack)callback taget:(NSObject*)taget;
{
 
    if(_eventsMap==nil){
        _eventsMap=[[NSMutableDictionary alloc]init];
    }
    NSMutableArray *list = [_eventsMap objectForKey:types];
    if(list==nil){
        list=[[NSMutableArray alloc]init];
        [_eventsMap setValue:list forKey:types];
    }
    EventBindVo* eventBindVo=[[EventBindVo alloc]init:callback b:taget];
    
    for (int i=0; i<list.count; i++) {
        EventBindVo* bin=list[i];
        if(bin.bfun==callback&&bin.thisObject==taget){
            return;
        }
    }
    [list addObject:eventBindVo];
    
 
}
-(void)removeEventListener:(NSString*)types  callback:(EventCallBack)callback taget:(NSObject*)taget;
{
    if(_eventsMap==nil){
        return;
    }
    NSMutableArray *list = [_eventsMap objectForKey:types];
    for (int i=0;list!=nil&& i<list.count; i++) {
        EventBindVo* bin=list[i];
        if(bin.bfun==callback&&bin.thisObject==taget){
            [list removeObjectAtIndex:i];
            return;
        }
    }
}
-(void)dispatchEvent:(BaseEvent*)event;
{
    if(_eventsMap==nil){
        return;
    }
    NSMutableArray *list = [_eventsMap objectForKey:event.type];
    for (int i=0;list!=nil&& i<list.count; i++) {
        EventBindVo* bin=list[i];
        bin.bfun(bin.thisObject,event);
  
 
    }
    
}
  
//public void dispatchEvent(BaseEvent event) {
//     HashMap<String, List<EventBindVo>> eventMap = this._eventsMap;
//     if (eventMap == null) {
//         return  ;
//     }
//     List<EventBindVo> list = eventMap.get(event.type);
//
//     if (list == null || list.size() == 0) {
//         return  ;
//     }
//     event.target = this;
//     for (int i = 0; i < list.size(); i++) {
//         EventBindVo eventBin = list.get(i);
//         eventBin.listener.call(eventBin.thisObject, event);
//     }
//
// }
 
@end

File: android/SceneWorld001/app/src/main/java/z3d/event/EventDispatcher.java
package z3d.event;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.function.Function;

public class EventDispatcher {
    protected HashMap<String, List<EventBindVo>>  _eventsMap ;

    public void addEventListener(String types   , EventCallBack listener, Object thisObject) {
        if (this._eventsMap==null) {
            this._eventsMap = new HashMap();
        }
        List<EventBindVo> list  = this._eventsMap.get(types);
        if (list==null) {
            this._eventsMap.put(types,new ArrayList());
            list  = this._eventsMap.get(types);
        }
        EventBindVo eventBin=new EventBindVo(listener,thisObject);
        for (int i= 0; i < list.size(); i++) {
            EventBindVo bin = list.get(i);
            if (bin.listener == listener && bin.thisObject == thisObject) {
                return;
            }
        }
        list.add(eventBin);

    }
    public void removeEventListener( String type,EventCallBack listener,Object thisObject) {
        if (this._eventsMap == null) {
            return;
        }
        List<EventBindVo> list = this._eventsMap.get(type);
        for (int i = 0; list!=null && i < list.size(); i++) {
            EventBindVo bin = list.get(i);
            if (bin.listener == listener && bin.thisObject == thisObject) {
//                list.splice(i, 1);
                list.remove(i);
                return;
            }
        }
    }
    public void dispatchEvent(BaseEvent event) {
        HashMap<String, List<EventBindVo>> eventMap = this._eventsMap;
        if (eventMap == null) {
            return  ;
        }
        List<EventBindVo> list = eventMap.get(event.type);

        if (list == null || list.size() == 0) {
            return  ;
        }
        event.target = this;
        for (int i = 0; i < list.size(); i++) {
            EventBindVo eventBin = list.get(i);
            eventBin.listener.call(eventBin.thisObject, event);
        }

    }


}
